plugin/trino-hive/src/main/java/io/trino/plugin/hive/CachingDirectoryLister.java

```diff
@@ -87,20 +95,69 @@ private static SchemaTablePrefix parseTableName(String tableName)
         }
         return new SchemaTablePrefix(schema, table);
     }
-
+    
+    /**
+     * Execute and return the result from {@link #listExecuter()} incorporated with a timer. 
+     * If listExecuter does not end within a certain amount of time (defined by 
+     * the field variable fileListingTimeout) the Thread running listExecuter will 
+     * be interrupted and a RuntimeException will be thrown. 
+     * 
+     * @param fs FileSystem, see: {@link org.apache.hadoop.fs.FileSystem}
+     * @param table Table, see: {@link io.trino.plugin.hive.metastore.Table}
+     * @param path  Path, see: {@link org.apache.hadoop.fs.Path}
+     * @return an iterator that runs over FileStatus-Object that inclues a file's block 
+     * locations
+     * @throws IOException
+     * @throws RuntimeException
+     */
     @Override
     public RemoteIterator<LocatedFileStatus> list(FileSystem fs, Table table, Path path)
-            throws IOException
-    {
+            throws IOException, RuntimeException {
+        ExecutorService timeoutExecutorService = Executors.newSingleThreadExecutor();
+        Future<RemoteIterator<LocatedFileStatus>> future = timeoutExecutorService
+                .submit(() -> listExecuter(fs, table, path));
+        try {
+            return future.get(fileListingTimeout.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            // something interrupted, probably your service is shutting down
+            timeoutExecutorService.shutdownNow();
+            throw new RuntimeException(e);
+        } catch (ExecutionException e) {
+            // error happened while executing listExecuter() - shut down this method
+            System.out.println(e);
+            timeoutExecutorService.shutdownNow();
+            throw new RuntimeException(e);
+        } catch (TimeoutException e) {
+            // timeout expired, stop this method from running and print an error message
+            future.cancel(true);
+            System.out.println(e);
```

Comment:
The praxis for printing errors is using `System.err.println()` instead of a regular println.